### ISSUE TITLE
Ported kindex file generator to GO. [phab:D1511]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1923]
_Created by **craigbarber** at 2017-04-11 23:43:05_: 

This was necessary due to the fact that the original script worked with TextFormat protos, running into a limitation when a multi-line input file was encountered. Namely, it would strip newlines from the input file, which causes downstream failures. After investing some time into debugging it, and consulting with fromberger, we decided it would be best to port the tool to Go so that it could work directly with binary protos to alleviate the problem.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-11 23:45:43_:

  - removed debug log from teest

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-11 23:46:38_:

  - removed extra space

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-11 23:47:43_:

  - split up long line

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-11 23:53:14_:

  - split up long comment line into multiple lines

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-11 23:56:50_:

  - split up long usage statement into multiple lines

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

The abbreviation "CPU" has its own separate meaning, and we don't use it for this anywhere else. "CU" would be fine for CompilationUnit, but better I think would be to spell it out:  

    signature = flag.String("signature", "", "The vname signature for the compilation unit")

or words to that effect

Actually, while I'm here: Do you really need to set the signature? It's not generally necessary for a compilation unit, since any node generated in its scope will be given a different one anyway. It's harmless to set one, but probably needn't be required.

For most units, the important pieces are language, corpus, (maybe) root, and (maybe) path.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

`output_file` or even just `output` would be sufficient here. Suggestion:

    outputPath = flag.String("output", "", "Path for output kindex file")

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

`source_file` or just `source`. As above.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

I'm not especially fond of the overly-terse messages this generates. It's better than nothing, but I think this would be better with an explicit usage:


```
func init() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, `Usage: %s -corpus <corpus> -language <language> -output <path> -source <path>

Construct a kindex file consisting of a single input file named by -source.
The resulting file is written to -output, and the vname of the compilation 
will be attributed to the specified -language and -corpus values.

Options:
`, filepath.Base(os.Args[0]))
		flag.PrintDefaults()
	}
}
```

It's a little more code, but it gives a more helpful message so the user doesn't have to go read the source code every time.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

Please call `flag.Parse()` at the beginning of `main()`. That's more conventional, and the absence of it there is surprising.

`func verifyFlags() { …`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

This isn't carrying its weight—please just use `if``, or a switch, e.g.,

```
	switch {
	case *vnameSignature == "":
		log.Fatal("You must provide a non-empty -vname_signature")
	case *vnamePath == "":
		log.Fatal("You must provide a non-empty -vname_path")
	case *corpus == "":
		log.Fatal("You must provide a non-empty -corpus label")
	case *language == "":
		log.Fatal("You must provide a non-empty -language label")
	case *outputPath == "":
		log.Fatal("You must provide a non-empty -output file path")
	case *sourcePath == "":
		log.Fatal("You must provide a non-empty -source path")
	}
```

Clarity > brevity, and in this case you actually get both from the simpler code. :)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

Another possible option here would be to let the user specify a Kythe URI, which carries corpus, language, path, root, and signature in a compact package. YMMV, but that would reduce the vname stuff to a single flag:

```
    unitURI = flag.String("uri", "", "A Kythe URI naming the compilation unit")

    …
    u, err := kytheuri.Parse(*unitURI)
    …
    // use u.Path, u.Corpus, u.Language, etc.
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

```
src, err := os.Open(*sourceFile)
if err != nil {
```

As a rule of thumb, try not to keep multiple error values alive in a scope unless you really need them to be (e.g., if you are checking two conditions separately after trying both). Re-using `err` is generally fine.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

```
if err := compilation.AddFile(*sourceFile, src, &spb.VName{
   …
}); err != nil {
   log.Fatalf(…)
}
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

```
out, err := os.Create(*outputPath)
if err != nil {
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-12 00:13:24_:

Don't ignore the error that can be returned by closing a //writable// file. Often errors are deferred till the close.

Suggestion here:

```
if _, err := compilation.WriteTo(out); err != nil {
   log.Fatalf("Error writing compilation to output: %v", err)
} else if err := out.Close(); err != nil {
   log.Fatalf("Error closing output file: %v", err)
}
```

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-12 17:34:34_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2017-04-12 17:36:53_

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-12 17:41:00_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** wrote at 2017-04-12 17:49:25_:

  - corrected example comment

**Comment by kythe** [comment]
_**craigbarber** closed via commit d746d1edad439af7de1aeba2eac6c3a71f761b4f at 2017-04-12 17:51:21_
